### PR TITLE
Added possibility to use FROM instructions at any time in Dockerfile …

### PR DESCRIPTION
**Pull request by thijslemmens**
…task. It will always end on top of the actual file

Looks to solve #338

### CODE REVIEW

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy [hunk 1]
```diff
@@ -31,23 +31,37 @@ class Dockerfile extends DefaultTask {
...
+                }
+            }
+        }
+    }
+
+    private Instruction findFromInstruction(){
+        Instruction fromInstruction
+        for(Instruction instruction: instructions){
+            if(instruction.keyword == 'FROM'){
+                fromInstruction = instruction
```
Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy]: We can just return at this point. No point in letting the for loop continue.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy [hunk 2]
```diff
@@ -31,23 +31,37 @@ class Dockerfile extends DefaultTask {
...
             }
         }
+        if(fromInstruction == null){
+            throw new IllegalStateException('Please specify a FROM instruction for your Dockerfile')
+        } else {
+            return fromInstruction
+        }
```
Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy]: If we return from inside the for loop we can just thrown the `IllegalStateException` if we reach this point.

**Comment by cdancy** [comment]
@thijslemmens couple minor comments to address but looks good so far.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy (hunk 1 above)
Comment (thijslemmens) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy]: My idea was using the last from instruction, making it possible to override. I added also some extra docs to the from methods for this.

Comment (cdancy) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy]: hmmm ... Ok I'll go with that. I can't imagine the use-case off the top of my head but it sounds reasonable to allow for this to happen.

Comment (thijslemmens) [inline src/main/groovy/com/bmuschko/gradle/docker/tasks/image/Dockerfile.groovy]: I do not have a use-case for it right now, but it was easy enough not to block it.

**Comment by thijslemmens** [comment]
Now I can see a problem with my own suggestion:
You can have multiple times FROM in one Dockerfile
https://docs.docker.com/engine/reference/builder/#/from states
FROM can appear multiple times within a single Dockerfile in order to create multiple images. Simply make a note of the last image ID output by the commit before each new FROM command.

I cannot imagine that I would want to use it, but it is something that my code does not support.

**Comment by cdancy** [comment]
@thijslemmens hmmm ... let me think on this. I've never actually built multiple images from a single Dockerfile and in fact did not know you code even do so ;)

**Comment by thijslemmens** [comment]
I asked a bit around, and I have not found anyone else that even knew of the multiple 'FROM' feature.
That said, I figured out that I could have solved the issue with the delayed from with a closure, more elegantly, so maybe, this PR is just an extra limitation.
That said, we could improve the error feedback to an earlier stage when using the from() methods and the instruction(String) method. In that case we can already detect weather the first instruction is a FROM or not. We can already fail in the pre-execution phases.

**Comment by cdancy** [comment]
@thijslemmens lets close this for now as you're able to move forward and the multiple FROM's options is still not yet clear to me how that would work given this context. Maybe best to just let things fail as they normally do.

**Pull request closed without merging**